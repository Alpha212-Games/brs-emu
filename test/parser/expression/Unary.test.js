const Parser = require("../../../lib/parser");
const Expr = require("../../../lib/parser/Expression");
const { Lexeme } = require("../../../lib/Lexeme");
const BrsError = require("../../../lib/Error");

const { token, EOF } = require("../ParserTests");

describe("parser", () => {
    afterEach(() => BrsError.reset());

    describe("unary expressions", () => {
        it("parses unary 'not'", () => {
            let parsed = Parser.parse([
                token(Lexeme.Not),
                token(Lexeme.True),
                EOF
            ]);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses consecutive unary 'not'", () => {
            let parsed = Parser.parse([
                token(Lexeme.Not),
                token(Lexeme.Not),
                token(Lexeme.Not),
                token(Lexeme.Not),
                token(Lexeme.Not),
                token(Lexeme.True),
                EOF
            ]);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses unary '-'", () => {
            let parsed = Parser.parse([
                token(Lexeme.Minus),
                token(Lexeme.Integer, 5),
                EOF
            ]);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses consecutive unary '-'", () => {
            let parsed = Parser.parse([
                token(Lexeme.Minus),
                token(Lexeme.Minus),
                token(Lexeme.Minus),
                token(Lexeme.Minus),
                token(Lexeme.Minus),
                token(Lexeme.Integer, 5),
                EOF
            ]);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });
    });
});
