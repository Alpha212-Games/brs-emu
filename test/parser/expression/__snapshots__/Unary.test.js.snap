// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser unary expressions parses consecutive unary '-' 1`] = `
Array [
  Expression {
    "expression": Unary {
      "operator": Object {
        "kind": 7,
        "line": 1,
        "literal": undefined,
      },
      "right": Unary {
        "operator": Object {
          "kind": 7,
          "line": 1,
          "literal": undefined,
        },
        "right": Unary {
          "operator": Object {
            "kind": 7,
            "line": 1,
            "literal": undefined,
          },
          "right": Unary {
            "operator": Object {
              "kind": 7,
              "line": 1,
              "literal": undefined,
            },
            "right": Unary {
              "operator": Object {
                "kind": 7,
                "line": 1,
                "literal": undefined,
              },
              "right": Literal {
                "value": 5,
              },
            },
          },
        },
      },
    },
  },
]
`;

exports[`parser unary expressions parses consecutive unary 'not' 1`] = `
Array [
  Expression {
    "expression": Unary {
      "operator": Object {
        "kind": 58,
        "line": 1,
        "literal": undefined,
      },
      "right": Unary {
        "operator": Object {
          "kind": 58,
          "line": 1,
          "literal": undefined,
        },
        "right": Unary {
          "operator": Object {
            "kind": 58,
            "line": 1,
            "literal": undefined,
          },
          "right": Unary {
            "operator": Object {
              "kind": 58,
              "line": 1,
              "literal": undefined,
            },
            "right": Unary {
              "operator": Object {
                "kind": 58,
                "line": 1,
                "literal": undefined,
              },
              "right": Literal {
                "value": BrsBoolean {
                  "kind": 1,
                  "value": true,
                },
              },
            },
          },
        },
      },
    },
  },
]
`;

exports[`parser unary expressions parses unary '-' 1`] = `
Array [
  Expression {
    "expression": Unary {
      "operator": Object {
        "kind": 7,
        "line": 1,
        "literal": undefined,
      },
      "right": Literal {
        "value": 5,
      },
    },
  },
]
`;

exports[`parser unary expressions parses unary 'not' 1`] = `
Array [
  Expression {
    "expression": Unary {
      "operator": Object {
        "kind": 58,
        "line": 1,
        "literal": undefined,
      },
      "right": Literal {
        "value": BrsBoolean {
          "kind": 1,
          "value": true,
        },
      },
    },
  },
]
`;
